src/lib.rs

```diff
@@ -166,12 +166,145 @@ pub fn broadcast<T>(cap: usize) -> (Sender<T>, Receiver<T>) {
     (s, r)
 }
 
+/// Create a new broadcast channel with the given limiter.
+pub fn broadcast_with_limiter<T, L>(capacity: L) -> (Sender<T, L>, Receiver<T, L>)
+where
+    L: CapacityLimiter<T>,
+{
+    let inner = Arc::new(Mutex::new(Inner {
+        queue: VecDeque::new(),
+        capacity,
+        overflow: false,
+        receiver_count: 1,
+        inactive_receiver_count: 0,
+        sender_count: 1,
+        head_pos: 0,
+        is_closed: false,
+        send_ops: Event::new(),
+        recv_ops: Event::new(),
+    }));
+
+    let s = Sender {
+        inner: inner.clone(),
+    };
+    let r = Receiver {
+        inner,
+        pos: 0,
+        listener: None,
+    };
+
+    (s, r)
+}
+
+/// A limit on the capacity of a channel.
+///
+/// This allows for more complex limits on a channel's capacity, beyond the number of elements.
+pub trait CapacityLimiter<T: ?Sized> {
+    /// An "over capacity" error.
+    type Error: error::Error;
+
+    /// Attempt to add the given item to the channel.
+    ///
+    /// Return `Ok` if the item may be added.  Any updates to internal state should only be done
+    /// if this function returns `Ok`.  An overflowing channel will retry the add after removing an
+    /// element if the returned error is not fatal.
+    fn try_add<'a, I>(&mut self, item: &T, contents: I) -> Result<(), Self::Error>
+    where
+        T: 'a,
+        I: ExactSizeIterator<Item = &'a T>;
+
+    /// Remove an item from the channel.
+    ///
+    /// The item removed and the rest of the items in the channel are provided in order to allow
+    /// updating internal state.  The default implementation does nothing.
+    fn remove<'a, I>(&mut self, item: &T, contents: I)
+    where
+        T: 'a,
+        I: ExactSizeIterator<Item = &'a T>,
+    {
+        let _ = (item, contents);
+    }
+
+    /// Hint if an error is fatal.
+    ///
+    /// Non-fatal errors will cause the channel to begin discarding elements (if the channel is
+    /// overflowing) or block on send (if not) until either the channel is empty or the `try_add`
+    /// succeeds.  If it is known that `try_add` will never succeed, return `true` here to cause
+    /// send operations to immediately return a `Rejected` error variant.
+    ///
+    /// The default implementation returns false.
+    fn error_is_fatal(&self, error: &Self::Error) -> bool {
```

Comment:
> It could be, but that would require another trait and add the burden of implementing another trait

Would it require that? You expanded the errors (e.g the `Rejected` variant) so it's obvious which variants are fatal and which are not?

Reply:
A non-fatal error is be "upgraded" to a fatal error if there are no messages to remove from the queue to make room.  This avoids an infinite loop (deadlock, not livelock) in the blocking sender.